filament/backend/src/vulkan/VulkanPipelineInfoCache.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_BACKEND_VULKANPIPELINEINFOCACHE_H
+#define TNT_FILAMENT_BACKEND_VULKANPIPELINEINFOCACHE_H
+
+#include <list>
+#include <vector>
+
+#include <bluevk/BlueVK.h>
+
+namespace filament::backend {
+
+struct VulkanVertexBuffer;
+
+// We use this class to cache pipeline metadata that can be computed out of performance-critical
+// sections such as draw() or beginRenderPass()
+class VulkanPipelineInfoCache {
+public:
+    struct VertexBufferInfo {
+        VulkanPipelineCache::VertexArray varray = {};
+        VkBuffer vkbuffers[MAX_VERTEX_ATTRIBUTE_COUNT] = {VK_NULL_HANDLE};
+        VkDeviceSize offsets[MAX_VERTEX_ATTRIBUTE_COUNT] = {0};
+        uint8_t attribToBufferIndex[MAX_VERTEX_ATTRIBUTE_COUNT] = {0};
+    };
+
+    VulkanPipelineInfoCache() {
+        increaseCapacity();
+    }
+
+    ~VulkanPipelineInfoCache() {
+        assert_invariant(mBuffers.size() == (mBacking.size() * CACHE_INCREMENT));
+    }
+
+    VertexBufferInfo* get() {
+        if (mBuffers.empty()) {
+            increaseCapacity();
+        }
+        auto ret = mBuffers.front();
+        mBuffers.pop_front();
+        return ret;
+    }
+
+    void release(VertexBufferInfo* buffer) {
+        mBuffers.push_back(buffer);
+    }
+
+private:
+    static constexpr size_t CACHE_INCREMENT = 10;
+
+    void increaseCapacity() {
+        mBacking.emplace_back(std::make_unique<VertexBufferInfoArray>());
+        auto& array = *(*(mBacking.end() - 1));
+        for (auto& info: array) {
+            mBuffers.push_back(&info);
+        }
+    }
+
+    using VertexBufferInfoArray = std::array<VertexBufferInfo, CACHE_INCREMENT>;
+    std::list<VertexBufferInfo*> mBuffers;
```

Comment:
`std::list` is pretty bad, it has 16-bytes overhead + it fragments the heap a lot, does lots of pointer chasing, etc... it's bad for the  cache. I wouldn't use. Honestly I would just use a `std::vector<VertexBufferInfo*>` instead. Also for something like this I wouldn't use `unique_ptr`, you should be able to do all the management yourself in the cache and avoid the overhead of unique_ptr.

You could also use a `std::deque` for `mBacking`, because objects don't move on insertion, so you could "simply" store a pointer in the `Handle<>`. But `deque` is really not efficient on some platforms (on windows).

I wonder if you couldn't do something more simple where you'd store a pointer in the `Handle<>` and it could point to a custom "pool" allocator. The size of the pool would have to be determined at compile time, but you can implement a "fallback" to malloc when it runs out of space + a log.

At the very least I would not use `std::list`, if insert/deletion is rare (which I assume it is), a simple `std::vector` will do. At least it'll fragment the heap much less.

Reply:
thanks for the heads-up on std::list.  I opted to just go with allocating on the heap for the struct, maybe we can revisit the pool in the future. Let me know if that's ok.